### PR TITLE
added dataprocessor , gitprovider and metricformator packages

### PR DESCRIPTION
**Pull request by Nikhil-dot-Kumar**
Checklist:

- [x] I have read [contribution guidelines](https://github.com/maplelabs/github-audit/blob/master/CONTRIBUTING.md).
- [x] My pull request has a descriptive title (not a vague title like `Update index.md`)
- [x] I have tested these changes locally on my machine.

##### SUMMARY

1. Added data processor package which process data from git APIs
2. Added gitprovider package which provides client for accessing vcs cloud provider APIs
3. Added metric formator package for customising metrics

##### COMPONENT NAME

[Data processor](https://github.com/maplelabs/github-audit/blob/master/docs/design/software_design.md#data-processor)
[Metric formator](https://github.com/maplelabs/github-audit/blob/master/docs/design/software_design.md#metric-formator)
[Gitprovider] : Provides utility to access cloud vcs APIs
##### ADDITIONAL INFORMATION

Sample metric formator json:
```json
{
    "changeDefaultKeys": {
        "document_type": "_documentType"
    },
    "addNewGlobalKeys": {
        "_plugin": "github-audit"
    }
}
```
changeDefaultKeys will replace output document keys with required customisation
addNewGlobalKeys will add new keys to every output document is required

### CODE REVIEW

File: gitprovider/gitprovider.go [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+// GitProvider represents git cloud provider which will be used to pull metrics
+type GitProvider interface {
+	// CheckCredentials checks user credentials
+	CheckCredentials() error
+
+	// GetCommits fetches commits using APIs
+	GetCommits(from time.Time, to time.Time, branch string) ([]byte, error)
+
+	// GetPullRequests fetches pull requests using APIs
+	GetPullRequests(int) ([]byte, error)
```
Comment (upendrasahu) [inline gitprovider/gitprovider.go:27]: what is int here?

File: internal/dataprocessor/githubprocessor.go [hunk 2]
```diff
@@ -0,0 +1,331 @@
...
+	var commits []github.RepositoryCommit
+	commitDocuments := make([]interface{}, 0)
+	err := json.Unmarshal(data, &commits)
+	if err != nil {
+		log.Errorf("error[%v] in unmarshalling commits", err)
+		return commitDocuments, err
+	}
+	for _, c := range commits {
+		var commit Commit
+		commit.DocumentType = "commit"
```
Comment (upendrasahu) [inline internal/dataprocessor/githubprocessor.go:219]: define var or const for different documentTypes

File: internal/dataprocessor/githubprocessor.go [hunk 3]
```diff
@@ -0,0 +1,331 @@
...
+		commit.Message = c.Commit.GetMessage()
+		commit.RepoType = "github"
```
Comment (upendrasahu) [inline internal/dataprocessor/githubprocessor.go:221]: define const for repotype

File: gitprovider/github.go [hunk 4]
```diff
@@ -0,0 +1,139 @@
...
+		}
+		opt.Page = resp.NextPage
+	}
+	allCommitsByte, err := json.Marshal(allCommits)
+	return allCommitsByte, err
+}
+
+// GetPullRequests fetches pull request for the user
+func (gc *GithubClient) GetPullRequests(fromNo int) ([]byte, error) {
+	opt := &github.PullRequestListOptions{
```
Comment (upendrasahu) [inline gitprovider/github.go:92]: Add debugs logs in all methods

File: gitprovider/gitprovider.go (hunk 1 above)
Comment (upendrasahu) [reply gitprovider/gitprovider.go:27]: add that var name as well

File: internal/dataprocessor/githubprocessor.go [hunk 5]
```diff
@@ -0,0 +1,353 @@
...
+	// Current time in milliseconds
+	CurrentTimeInMS int64
+
+	// Metricformator instance to customise processed data
+	MetricFormator *metricformator.MetricFormator
+}
+
+// NewGithubProcessor provides new instance of github api processor
+func NewGithubProcessor(repoName string, repoURL string) GithubProcessor {
+	var gp GithubProcessor
```
Comment (upendrasahu) [inline internal/dataprocessor/githubprocessor.go:36]: Instead of creating new var, u can simply do 
return GithubProcessor{RepoName: repoName, ...

Comment (Nikhil-dot-Kumar) [reply internal/dataprocessor/githubprocessor.go:36]: defining other params also , for time variable etc so added as separate variable declaration

**Pull request merged**